graylog2-server/src/main/java/org/graylog/mcp/server/Tool.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.mcp.server;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.joda.JodaModule;
+import com.github.victools.jsonschema.generator.OptionPreset;
+import com.github.victools.jsonschema.generator.SchemaGenerator;
+import com.github.victools.jsonschema.generator.SchemaGeneratorConfig;
+import com.github.victools.jsonschema.generator.SchemaGeneratorConfigBuilder;
+import com.github.victools.jsonschema.generator.SchemaVersion;
+import com.github.victools.jsonschema.module.jackson.JacksonModule;
+import com.github.victools.jsonschema.module.jackson.JacksonOption;
+import org.graylog.jsonschema.EmptyObjectAsObjectModule;
+import org.graylog.mcp.tools.PermissionHelper;
+import org.graylog2.plugin.security.Permission;
+
+import java.util.Map;
+
+/**
+ * The base class for MCP tools.
+ *
+ * @param <P> Parameter type
+ * @param <O> Output type
+ */
+public abstract class Tool<P, O> {
+
+    // MCP inexplicable uses Draft 7 of JSON Schema
+    private static final SchemaGeneratorConfigBuilder CONFIG_BUILDER = new SchemaGeneratorConfigBuilder(SchemaVersion.DRAFT_7, OptionPreset.PLAIN_JSON);
+    private static final SchemaGeneratorConfig CONFIG = CONFIG_BUILDER.with(new EmptyObjectAsObjectModule())
+            .with(new JacksonModule(JacksonOption.RESPECT_JSONPROPERTY_REQUIRED)).build();
+    private static final SchemaGenerator GENERATOR = new SchemaGenerator(CONFIG);
+
+    private final ObjectMapper objectMapper;
+    private final TypeReference<P> parameterType;
+    private final String name;
+    private final String title;
+    private final String description;
+    private final String inputSchema;
+
+    protected Tool(ObjectMapper objectMapper, TypeReference<P> parameterType, String name, String title, String description) {
+        this.objectMapper = objectMapper;
+        this.objectMapper.registerModule(new JodaModule());
```

Comment:
We should not modify the object mapper as this is a shared instance.
in addition the JodaModule is already installed.